sdk/src/test/scala/com/horizen/account/state/CertificateKeyRotationMsgProcessorTest.scala

```diff
@@ -0,0 +1,191 @@
+      assertEquals("Certifiers signing keys array is incorrect", oldSigningKey.map(_.publicImage()), certifiersKeys.signingKeys)
+
+    }
+  }
+
+  @Test
+  def testKeyRotationOverride(): Unit = {
+    val oldMasterKey = generateKeys(1)
+    val oldSigningKey = generateKeys(1)
+    val newMasterKeyFirst = generateKeys(1).head
+    val newMasterKeySecond = generateKeys(1).head
+
+    val keyRotationProofFirst = KeyRotationProof(
+      masterKeyType,
+      0,
+      newMasterKeyFirst.publicImage(),
+      oldSigningKey.head.sign(newMasterKeyFirst.getPublicBytes.take(32)),
+      oldMasterKey.head.sign(newMasterKeyFirst.getPublicBytes.take(32))
+    )
+    val keyRotationProofSecond = KeyRotationProof(
+      masterKeyType,
+      0,
+      newMasterKeySecond.publicImage(),
+      oldSigningKey.head.sign(newMasterKeySecond.getPublicBytes.take(32)),
+      oldMasterKey.head.sign(newMasterKeySecond.getPublicBytes.take(32))
+    )
+
+    usingView(certificateKeyRotationMsgProcessor) { view =>
+
+      certificateKeyRotationMsgProcessor.init(view)
+      when(mockNetworkParams.signersPublicKeys).thenReturn(oldSigningKey.map(_.publicImage()))
+      when(mockNetworkParams.masterPublicKeys).thenReturn(oldMasterKey.map(_.publicImage()))
+
+      withGas {
+        certificateKeyRotationMsgProcessor.process(
+          getMessage(
+            to = contractAddress,
+            data = BytesUtils.fromHexString(SubmitKeyRotationReqCmdSig) ++
+              SubmitKeyRotationCmdInput(keyRotationProofFirst, newMasterKeyFirst.sign(newMasterKeyFirst.getPublicBytes.take(32))).encode(),
+            nonce = randomNonce
+          ), view, _, defaultBlockContext
+        )
+      }
+      withGas {
+        certificateKeyRotationMsgProcessor.process(
+          getMessage(
+            to = contractAddress,
+            data = BytesUtils.fromHexString(SubmitKeyRotationReqCmdSig) ++
+              SubmitKeyRotationCmdInput(keyRotationProofSecond, newMasterKeySecond.sign(newMasterKeySecond.getPublicBytes.take(32))).encode(),
+            nonce = randomNonce
+          ), view, _, defaultBlockContext
+        )
+      }
+
+      val masterKeyHistory = certificateKeyRotationMsgProcessor.getKeysRotationHistory(masterKeyType, 0, view)
+      assertEquals("Key Rotation history is not updated", List(0), masterKeyHistory.epochNumbers)
+
+      val signingKeyHistory = certificateKeyRotationMsgProcessor.getKeysRotationHistory(signingKeyType, 0, view)
+      assertEquals("Key Rotation history was mistakenly updated", true, signingKeyHistory.epochNumbers.isEmpty)
+
+      val maybeKeyRotation = certificateKeyRotationMsgProcessor.getKeyRotationProof(masterKeyType, 0, 0, view)
+      assertTrue("Key rotation proof is not persisted", maybeKeyRotation.isDefined)
+      assertEquals("Key rotation proof is corrupted", keyRotationProofSecond, maybeKeyRotation.get)
+
+    }
```

Comment:
It would be nice to add another test where the key is changed in 2 different epochs

Reply:
Added test for the next scenario:
```
    SETUP:
    config with 2 signing keys and 2 master keys
    master key #0 is changed in epoch 0, 1, 2
    signing key #0 is changed in epoch 1
    master key #1 is changed in epoch 2
    signing key #1 is never changed

    VERIFICATIONS:
    for epoch 0 all keys from config
    for epoch 1 master key #0 is changed
    for epoch 2 master key #0 is changed twice and signing key #1 is changed
    ```